### PR TITLE
Cleanup Queues, Part 1: Schema Definition

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Part 1 of building a cleanup queue, to be used by Sweep for async cleanup tasks, and possibly later on for other internal uses e.g. targeted sweep.

**Implementation Description (bullets)**:
- Implement a generator for queue-like tables.
- Implement a Cleanup Schema, which uses this generator in a pretty bare form.

**Concerns (what feedback would you like?)**:
- The lines in `getSequentialTableDefinition` probably need to be reviewed very closely, as there are a number of trade-offs and assumptions I've made there.
- We're using blobs for the row key and values. I intend them to be this way so that I can write arbitrary things there (Part 2 will include a protobuf for a cell batch).

**Where should we start reviewing?**:`SequentialTables`, then `CleanupSchema`. The rest is generated.

**Priority (whenever / two weeks / yesterday)**: one week

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2768)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/CleanupSchema.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+package com.palantir.atlasdb.schema;
+
+import java.io.File;
+
+import com.palantir.atlasdb.keyvalue.api.Namespace;
+import com.palantir.atlasdb.schema.queue.SequentialTables;
+import com.palantir.atlasdb.table.description.OptionalType;
+import com.palantir.atlasdb.table.description.Schema;
+
+public enum CleanupSchema implements AtlasSchema {
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/CleanupSchema.java]: can we name this a bit more descriptively?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    private static TableDefinition getSequentialTableDefinition() {
+        return new TableDefinition() {{
+            allSafeForLoggingByDefault();
+            conflictHandler(ConflictHandler.IGNORE_ALL); // Offsets are unique
+            rowName();
+                hashFirstRowComponent(); // Can still find all queues by virtue of the offset table
+                rowComponent("queue_key", ValueType.BLOB);
+            dynamicColumns();
+                columnComponent("offset", ValueType.VAR_LONG);
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java]: we talked about this in person so you may be addressing this already - but we want to partition the rows to limit the number of columns in any given row

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        schema.addTableDefinition(sequentialTablePrefix + "_write_offsets", getOffsetTableDefinition());
+    }
+
+    private static TableDefinition getSequentialTableDefinition() {
+        return new TableDefinition() {{
+            allSafeForLoggingByDefault();
+            conflictHandler(ConflictHandler.IGNORE_ALL); // Offsets are unique
+            rowName();
+                hashFirstRowComponent(); // Can still find all queues by virtue of the offset table
+                rowComponent("queue_key", ValueType.BLOB);
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java]: will this always be the table name? (seems like it given that the offsets table key is `full_table_name`)

**Comment by nziebart** [comment]
@tpetracca for SA, since he is working on an internal impl of the queue table using dynamic columns

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/CleanupSchema.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/schema/CleanupSchema.java]: Yep, I'll update this to `CleanupTaskQueueSchema`, or something along those lines if I can think of something better.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java]: This was actually intended to be a pair of the table name and an ID (for the partitioning we discussed). I think this is an artifact of me trying to allow queue keys of completely arbitrary types.

Really we should go for something like
```
hashFirstRowComponents(2);
rowComponent("full_table_name", STRING);
rowComponent("row_partition", VAR_LONG);
```

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/schema/queue/SequentialTables.java]: (Yep, indeed - as mentioned above, there should be 2 distinct row components for table-name and the partition.)

**Comment by tboam** [comment]
@jeremyk-91 to address comments, check tests and update changelog

**Comment by jeremyk-91** [comment]
I'm going to close this out in the short run. We'll have use for this eventually, perhaps as part of support with the internal storage-managing product. However, we're going with a different solution for Nimbus for now.

**Pull request closed without merging**